### PR TITLE
Add support for ActiveMarkings to DocumentTraverser

### PR DESCRIPTION
**Pull request by BrettMichaelJohnson**

### CODE REVIEW

File: src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java [hunk 1]
```diff
@@ -317,11 +319,29 @@ private boolean hasPermissions(Permissions.Acl permissions,
...
+        ? Collections.emptyIterator() : activeMarkings.iterator();
+    while (iterator.hasNext()) {
+      Marking marking = ((ActiveMarking) iterator.next()).get_Marking();
+      Permissions.Acl markingPerms =
+          new Permissions(marking.get_Permissions()).getAcl();
+      Acl markingAcl = createAcl(docId, secParentFragment,
+          Acl.InheritanceType.AND_BOTH_PERMIT, options.getGlobalNamespace(),
+          markingPerms.getAllowUsers(), markingPerms.getDenyUsers(),
+          markingPerms.getAllowGroups(), markingPerms.getDenyGroups());
+      if (markingAcl != null) {
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java]: Why not log absent active markings here? Maybe re-unify these code blocks in the followup CL (setting secParentFragment first everywhere, etc.)?

File: src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java [hunk 2]
```diff
@@ -348,25 +368,23 @@ private void processInheritedPermissions(DocId docId,
...
-        && allowGroups.isEmpty() && denyGroups.isEmpty()) {
-      return null;
-    } else {
-      return createAcl(docId, parentFragment,
-          Acl.InheritanceType.CHILD_OVERRIDES,
-          options.getGlobalNamespace(), allowUsers, denyUsers,
-          allowGroups, denyGroups);
-    }
+    return createAcl(docId, parentFragment,
+        Acl.InheritanceType.CHILD_OVERRIDES,
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java]: I think I would just have added the inheritance type as a parameter here and used it for all the inherited ACLs. If you just didn't want to touch the existing code, maybe do that in the followup, too?

File: src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java [hunk 3]
```diff
@@ -348,25 +368,23 @@ private void processInheritedPermissions(DocId docId,
...
+        permissions.getDenyGroups(permSrc));
   }
 
   private Acl createAcl(DocId docId, String parentFragment,
       Acl.InheritanceType inheritanceType, String namespace,
       Set<String> allowUsers, Set<String> denyUsers,
       Set<String> allowGroups, Set<String> denyGroups) {
+    if (allowUsers.isEmpty() && denyUsers.isEmpty()
+        && allowGroups.isEmpty() && denyGroups.isEmpty()) {
+      return null;
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java]: I don't think this is right for marking sets or documents. For marking sets, no principals in the permissions means no access, not a null ACL, and for documents, we still need to set the inherit-from.

File: test/com/google/enterprise/adaptor/filenet/DocumentTraverserTest.java [hunk 4]
```diff
@@ -356,6 +359,80 @@ public void testGetDocContent() throws Exception {
...
+
+    acl = response.getNamedResources().get(markingFrag(markingId1));
+    assertFalse(acl.getPermitUsers().toString(),
+        acl.getPermitUsers().isEmpty());
+    assertEquals(InheritanceType.AND_BOTH_PERMIT, acl.getInheritanceType());
+    assertEquals(null, acl.getInheritFrom());
+    assertEquals(null, acl.getInheritFromFragment());
+  }
+
+  private String markingFrag(String id) {
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/filenet/DocumentTraverserTest.java]: s/Frag/Fragment/

File: test/com/google/enterprise/adaptor/filenet/DocumentTraverserTest.java [hunk 5]
```diff
@@ -356,6 +359,80 @@ public void testGetDocContent() throws Exception {
...
+    return "MARK" + id.replace("{", "%7B").replace("}", "%7D");
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/filenet/DocumentTraverserTest.java]: We have this code twice in DocumentTraverser now, too, so maybe add a helper method like 

String percentEscape(Id)

File: src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java (hunk 2 above)
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java]: That doesn't work.  The only cases that can and do use this method only need CHILD_OVERRIDES.
The LEAF node passes in a union for each set.  The markings do not have a PermissionSource.

File: test/com/google/enterprise/adaptor/filenet/DocumentTraverserTest.java (hunk 4 above)
Comment (BrettMichaelJohnson) [inline test/com/google/enterprise/adaptor/filenet/DocumentTraverserTest.java]: Done.

File: src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java (hunk 1 above)
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java]: Added marking ACL logging.

File: src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java [hunk 6]
```diff
@@ -322,69 +327,56 @@ private void processInheritedPermissions(DocId docId,
       ActiveMarkingList activeMarkings, Permissions.Acl permissions,
       Response response) {
     // Send add request for adding ACLs inherited from parent folders.
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java]: Er, move this comment down? Add a comment here that active marking sets must be ANDed with the other permissions, and the GSA only supports AND_BOTH_PERMIT (at least, with the correct semantics) at the root of the inheritance chain.

File: src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java [hunk 7]
```diff
@@ -317,56 +324,59 @@ private boolean hasPermissions(Permissions.Acl permissions,
...
-    if (folderAcl == null) {
-      logger.log(Level.FINEST,
-          "{0} does not have inherited folder permissions", docId);
-    } else {
-      logger.log(Level.FINEST, "Create ACL for folder {0}#{1}: {2}",
-          new Object[] {docId, SEC_FOLDER_POSTFIX, folderAcl});
-      response.putNamedResource(SEC_FOLDER_POSTFIX, folderAcl);
-      secParentFragment = SEC_FOLDER_POSTFIX;
+    String fragment = null;
+    Iterator<?> iterator = (activeMarkings == null)
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java]: This can change to avoid the null check when you resolve the conflicts.

**Pull request merged**